nntrainer/layers/lstm.cpp

```diff
@@ -218,11 +223,118 @@ void LSTMLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void LSTMLayer::calcDerivative() {
-  // NYI
+  Tensor &derivative_ = fgio->getGradientRef();
+  Tensor &weight =
+    weightAt(static_cast<int>(LSTMParams::weight_xh)).getVariableRef();
+  Tensor &ret_ = net_input[0]->getGradientRef();
+  derivative_.dot(weight, ret_, false, true);
 }
 
 void LSTMLayer::calcGradient() {
-  // NYI
+  Tensor &djdw_x =
+    weightAt(static_cast<int>(LSTMParams::weight_xh)).getGradientRef();
+  Tensor &djdw_h =
+    weightAt(static_cast<int>(LSTMParams::weight_hh)).getGradientRef();
+  Tensor &djdb_h =
+    weightAt(static_cast<int>(LSTMParams::bias_h)).getGradientRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(LSTMParams::weight_hh)).getVariableRef();
+
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
```

Comment:
consider updating net_hidden to net_output someday as now it confuses with the hidden state of RNN and LSTM :).

Reply:
yes. it is. but actually they are the same.